tfhe/src/keycache/mod.rs

```diff
@@ -186,32 +200,44 @@ pub mod utils {
                         .map(|param_key| param_key.1.clone())
                 };
 
-                if let Some(shared_cell) = maybe_shared_cell {
-                    shared_cell.inner.get_or_init(load_from_persistent_storage);
-                    Ok(shared_cell)
-                } else {
-                    Err(())
-                }
+                maybe_shared_cell.map_or_else(
+                    || {
+                        let has_found_key = {
+                            self.memory_storage
+                                .read()
+                                .unwrap()
+                                .iter()
+                                .any(|(p, _)| *p == param)
+                        };
+
+                        if !has_found_key {
+                            let shared_key = SharedKey {
+                                inner: Arc::new(OnceLock::new()),
+                            };
+                            shared_key.inner.get_or_init(load_from_persistent_storage);
+                            {
+                                // we only hold a write lock for a short duration to push the lazily
+                                // evaluated key without actually evaluating the key
+                                let mut memory_storage = self.memory_storage.write().unwrap();
+                                if !memory_storage.iter().any(|(p, _)| *p == param) {
```

Comment:
I think I prefer

```rust
if memory_storage.iter().all(|(p, _)| *p != param)
```

i.e. if all available params are not my params then it is missing the key

Reply:
It seems not right, in that case it would means that all keys stored in the directory must have the same parameters set in order for this conditional test to pass no ?

Reply:
we check that all parameters in the vector are not equal to our parameters, which means our parameters are not in the vector, if that's the case we add our parameters + key we just generated

Reply:
it's a matter of taste, both conditions should be equivalent, I just find my formulation easier to read, but I guess it varies from person to person :)

Reply:
Oh yes, I miss the *not* equal part in my head. So your proposal makes more sense and is definitely clearer.